core/codegen/src/attribute/route.rs

```diff
@@ -414,6 +429,25 @@ fn codegen_route(route: Route) -> Result<TokenStream> {
     let path = route.attribute.path.origin.0.to_string();
     let rank = Optional(route.attribute.rank);
     let format = Optional(route.attribute.format);
+    let mut states: Vec<Option<&syn::Type>> = vec![None; 16];
+
+    // Find all managed states
+    let mut iter = 0;
+    for (_arg_ident, _, ty) in &route.inputs {
+        match ty {
+            syn::Type::Path(type_path) => {
+                if let Some(path_segment) = type_path.path.segments.first() {
+                    if path_segment.ident == "State" {
```

Comment:
This is the most fragile part of the implementation, since it assumes the `State` type is actually named `State`. There might be some type-level tricks we could employ, but it would probably be pretty convoluted.

Assuming we only match by exact name (which I am not super happy about), this should be checking that the *last* segment (before generic arguments) is `State`.

Reply:
Yeah, I'm worried about this as well, but I couldn't think of any other way to compare the types.

Reply:
Here's a totally untested concept, but I think the `impl` rules allow this:

```rust
// Usual "do not use this outside of Rocket" warning comment
#[doc(hidden)]
trait GetStateTypeId {
    const fn get_typeid() -> Option<TypeId>;
}

impl<T> GetStateTypeId for T {
    const fn get_typeid() -> Option<TypeId> { None }
}

impl<T> GetStateTypeId for State<T> {
    const fn get_typeid() -> Option<TypeId> { Some(type_id) }
}
```

Then emit code like `state_types: &[#( <#ty as #GetStateTypeId>::get_typeid() )]`. This is a bit inefficient to work with because of `None`, but those checks should only happen once at launch.

Reply:
Unfortunately that would require [specialization (RFC 1210)](https://github.com/rust-lang/rfcs/blob/master/text/1210-impl-specialization.md), which was marked as not needed any more in Rocket (https://github.com/SergioBenitez/Rocket/issues/19):
` Update (Sep 9, 2019): specialization is no longer used.`
and I'd love to keep it that way because I don't know if there is a timeline for this RFC but it seems pretty convoluted.

I found this article which describes a code that achieves specialization by using autoref in a specific way: https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md
I wrote this code that implements it for `State`, but I haven't yet had time to think about integration with this PR:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4b2bceca8dbe0e744aee5af4080b3628

Reply:
That's disappointing; I misremembered and thought that what I posted didn't need specialization. But anyway it still only solves `State` itself and not state used internally in `FromRequest` impls; maybe a better solution exists that solves both.